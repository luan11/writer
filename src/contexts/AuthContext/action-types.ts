const prefix = 'AuthContext';

export const LOGIN = `${prefix}/LOGIN`;
export const LOGOUT = `${prefix}/LOGOUT`;
export const SET_LOADING = `${prefix}/SET_LOADING`;
export const SET_LOADED = `${prefix}/SET_LOADED`;