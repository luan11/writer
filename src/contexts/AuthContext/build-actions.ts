import * as actionTypes from './action-types';

type DispatchProps = {
  type: string;
  payload?: {};
};

type Dispatch = (props: DispatchProps) => void;

export function buildActions(dispatch: Dispatch) {
  return {
    login: (payload: {}) => dispatch({ type: actionTypes.LOGIN, payload }),
    logout: () => dispatch({ type: actionTypes.LOGOUT }),
    setLoading: () => dispatch({ type: actionTypes.SET_LOADING }),
    setLoaded: () => dispatch({ type: actionTypes.SET_LOADED }),
  };
}