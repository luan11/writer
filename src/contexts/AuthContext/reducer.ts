import { AuthContextData } from '.';

import * as actionTypes from './action-types';

type ActionProps = {
  type: string;
  payload?: {};
};

export function reducer(state: AuthContextData, action: ActionProps) {
  switch(action.type) {
    case actionTypes.LOGIN:
      return { ...state, ...action.payload };

    case actionTypes.LOGOUT:
      return { ...state, token: null, authenticated: false }

    case actionTypes.SET_LOADING:
      return { ...state, loading: true }

    case actionTypes.SET_LOADED:
      return { ...state, loading: false }
  }

  return state;
}