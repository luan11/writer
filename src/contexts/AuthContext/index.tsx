import { createContext, ReactNode, useEffect, useReducer, useRef } from 'react';

import { buildActions, LoginProps } from './build-actions';
import { reducer } from './reducer';

import api from './../../services/api';

import { TOKEN_KEY, getToken } from './../../utils/auth';

const token = getToken();

export type AuthContextData = {
  token: string | null;
  authenticated: boolean;
  loading: boolean;
};

export const initialState: AuthContextData = {
  token: token,
  authenticated: token ? true : false,
  loading: false
};

type Actions = {
  login: (payload: LoginProps) => void;
  logout: () => void;
  setLoading: () => void;
  setLoaded: () => void;
};

type AuthContextProps = {
  state: AuthContextData;
  actions: Actions;
  doLogin: (payload: LoginProps) => void;
  doLogout: () => void;
};

export const AuthContext = createContext({} as AuthContextProps);

type AuthContextProviderProps = {
  children: ReactNode;
};

export function AuthContextProvider({ children }: AuthContextProviderProps) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const actions = useRef(buildActions(dispatch));

  const { token } = state;

  useEffect(() => {
    api.defaults.headers['Authorization'] = `Bearer ${token}`;
  }, [token]);

  function doLogin(payload: LoginProps) {
    localStorage.setItem(TOKEN_KEY, payload.token);

    actions.current.login(payload);
  }

  function doLogout() {
    localStorage.removeItem(TOKEN_KEY);

    actions.current.logout();
  }

  return (
    <AuthContext.Provider value={{
      state,
      actions: actions.current,
      doLogin,
      doLogout,
    }}>
      {children}
    </AuthContext.Provider>
  );
}