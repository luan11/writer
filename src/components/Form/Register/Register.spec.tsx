import { render, screen } from '@testing-library/react';

import { AuthContextProvider } from './../../../contexts/AuthContext';

import { Form } from '.';

describe('<Register />', () => {
  it('should render Register correctly', () => {
    render(
      <AuthContextProvider>
        <Form />
      </AuthContextProvider>
    );

    expect(screen.getByTestId('username')).toBeInTheDocument();
    expect(screen.getByTestId('password')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <AuthContextProvider>
        <Form />
      </AuthContextProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});