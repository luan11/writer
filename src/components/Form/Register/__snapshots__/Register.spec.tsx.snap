// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> should match snapshot 1`] = `
<form
  autocomplete="off"
  class="sc-bdnxRM fvJqbn"
>
  <h2
    class="sc-gtsrHT gNhmEX"
  >
    Crie sua conta
  </h2>
  <div
    class="sc-dlnjwi ctuddm"
  >
    <label
      class="sc-hKFxyN dCXiQB"
    >
      Usu√°rio
    </label>
    <input
      class="sc-eCApnc cqYdUq"
      data-testid="username"
      name="username"
      placeholder=" "
      type="text"
    />
  </div>
  <div
    class="sc-dlnjwi ctuddm"
  >
    <label
      class="sc-hKFxyN dCXiQB"
    >
      Senha
    </label>
    <input
      class="sc-eCApnc cqYdUq"
      data-testid="password"
      name="password"
      placeholder=" "
      type="password"
    />
  </div>
  <button
    class="btn"
    type="submit"
  >
    Criar
  </button>
</form>
`;
